pandas/_libs/tslibs/offsets.pyx

```diff
@@ -180,7 +180,7 @@ def apply_wraps(func):
             result = result.normalize()
 
         # nanosecond may be deleted depending on offset process
-        if not self.normalize and nano != 0:
+        if not self.normalize and nano != 0 and not hasattr(self, "nanoseconds"):
```

Comment:
I don't think so, the RelativeDelta can already add the nano seconds correctly, but this wraper is trying to add the nanoseconds again, so the codes of wrapper should be changed. Otherwise, we could implement another wrapper, which does not make too much sense.

Or could you suggest a solution?

Reply:
why is this 'nanoseconds' and not 'nanosecond'? can this be a property on the offsets ?

Reply:
e.g L168 looks odd compared to this

Reply:
> why is this 'nanoseconds' and not 'nanosecond'? can this be a property on the offsets ?

It could be either `nanoseconds` or `nanosecond`. Please look at the codes below:
```Python
from pandas.tseries.offsets import DateOffset

>>> DateOffset(nanosecond=1).__dict__
{'_offset': datetime.timedelta(1), '_use_relativedelta': False, 'nanosecond': 1}

>>> DateOffset(nanoseconds=1).__dict__
{'_offset': datetime.timedelta(1), '_use_relativedelta': False, 'nanoseconds': 1}
```

Reply:
But `pandas.Timestamp` only accept `nanosecond` as parameter. E.g.
```Python
from pandas import Timestamp

>>> Timestamp(year=2000, month=1, day=1,  nanosecond=0)
Timestamp('2000-01-01 00:00:00')

>>> Timestamp(year=2000, month=1, day=1,  nanoseconds=0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/tslibs/timestamps.pyx", line 936, in pandas._libs.tslibs.timestamps.Timestamp.__new__
TypeError: __new__() got an unexpected keyword argument 'nanoseconds'
```

Reply:
So actually, `RelativeDeltaOffset` cannot add/sub correctly either with `nanoseconds` or `nanosecond`. We need to handle both, I guess.

Reply:
no not true, from the doc-string

```
    Parameters that **add** to the offset (like Timedelta):

    - years
    - months
    - weeks
    - days
    - hours
    - minutes
    - seconds
    - microseconds
    - nanoseconds

    Parameters that **replace** the offset value:

    - year
    - month
    - day
    - weekday
    - hour
    - minute
    - second
    - microsecond
    - nanosecond.
```

Reply:
Hey @jreback ,
Thanks for explaining this to me. Let's come back to why `hasattr(self, "nanoseconds")` is needed.
Because line 168 `nano = other.nanosecond` and line 184 `if result.nanosecond != nano:` do not make sense here.

This statement assumes the `result.nanosecond` is only affected by `other.nanosecond`(Timestamp), but actually`self.nanoseconds` can also affect the `result.nanosecond`.

For instance,
```Python
t = Timestamp(year=2000, month=12, day=1, nanosecond=3) + DateOffset(nanoseconds=7)
```
L175: We get a correct result already.
L168: `nano` would be 3,
L184:`result.nanosecond` would be 10
L184: `if result.nanosecond != nano` (10 != 3) would be True.
L190:  We are adding the correct result another 3 nanoseconds.

Finally, we get a wrong result.

So what is your advice here, thanks~

Reply:
@jreback 

Another question is: what behavior do we expect for the example below?
```Python
Timestamp(year=2000, month=12, day=1, nanosecond=3) + DateOffset(nanoseconds=7, nanosecond=1)
```
Add first then replace, or replace first then add?
Shall we raise an error when users trying to do this?

Reply:
we should raise if we get a singluar & plural form (don't do it in this PR), do a separate one.

Reply:
Thanks and could you also response to my previous reply for why `hasattr(self, "nanoseconds")` is needed?

Reply:
> Add first then replace, or replace first then add?  Shall we raise an error when users trying to do this?

I think relativedelta does replace-then-add, but we should double-check (and match its behavior)

Reply:
@tushushu this doesn't make sense here. we *cannot* have the plural form at all.

Reply:
> Hey @jreback , Thanks for explaining this to me. Let's come back to why `hasattr(self, "nanoseconds")` is needed. Because line 168 `nano = other.nanosecond` and line 184 `if result.nanosecond != nano:` do not make sense here.
> 
> This statement assumes the `result.nanosecond` is only affected by `other.nanosecond`(Timestamp), but actually`self.nanoseconds` can also affect the `result.nanosecond`.
> 
> For instance,
> 
> ```python
> t = Timestamp(year=2000, month=12, day=1, nanosecond=3) + DateOffset(nanoseconds=7)
> ```
> 
> L175: We get a correct result already. L168: `nano` would be 3, L184:`result.nanosecond` would be 10 L184: `if result.nanosecond != nano` (10 != 3) would be True. L190: We are adding the correct result another 3 nanoseconds.
> 
> Finally, we get a wrong result.
> 
> So what is your advice here, thanks~

@jreback Thanks for your reply. Anyway we need this line to handle case like `t = Timestamp(year=2000, month=12, day=1, nanosecond=3) + DateOffset(nanoseconds=7)`.

Reply:
> this doesn't make sense here. we _cannot_ have the plural form at all

@jreback i dont why you claim this (and have done so elsewhere xref #7418).  RelativeDeltaOffset behaves like dateutil's `relativedelta`, in which `obj.second` and `obj.seconds` mean different things.

Reply:
@jreback Could you please respond me and @jbrockmendel?

Reply:
of course
my point is that o don't why you think should be nanoseconds and not nanosecond

Reply:
> 

Because the `self` in that line is an instance of `RelativeDeltaOffset`, and the title of this PR shows we are solving the issue of `RelativeDeltaOffset` with `nanoseconds` cannot be added to `Timestamp` correctly, but not `nanosecond`. 

If we want to fix the behavior of `RelativeDeltaOffset` with `nanosecond` add to `Timestamp`, we can do it in another PR, I guess.